### ISSUE TITLE
How to convert the model to onnx in libtorch? 

### ISSUE THREAD

**Issue by bjliuzp** [pytorch/pytorch#33343]
struct Net : torch::nn::Module {
	Net()
		: conv1(torch::nn::Conv2dOptions(1, 20, /*kernel_size=*/5).stride(1)),
		conv2(torch::nn::Conv2dOptions(20, 40, /*kernel_size=*/5)),
		fc1(640, 120),
		fc2(120, 10) {
		register_module("conv1", conv1);
		register_module("conv2", conv2);
		register_module("conv2_drop", conv2_drop);
		register_module("fc1", fc1);
		register_module("fc2", fc2);
	}
	torch::Tensor forward(torch::Tensor x) {
		x = torch::relu(torch::max_pool2d(conv1->forward(x), 2));//(28-5)+1=24,12 x 12 x 10
		x = torch::relu(torch::max_pool2d(conv2_drop->forward(conv2->forward(x)), 2));//(12-5)+1=8,4 x 4 x 20
		//x = torch::relu(torch::avg_pool2d(conv2_drop->forward(conv2->forward(x)), 2));//(12-5)+1=8,4 x 4 x 20

		x = x.view({ -1, 640 });
		x = torch::relu(fc1->forward(x));
		x = torch::dropout(x, /*p=*/0.5, /*training=*/is_training());
		x = fc2->forward(x);
		return torch::log_softmax(x, /*dim=*/1);
	}
	torch::nn::Conv2d conv1;
	torch::nn::Conv2d conv2;
	torch::nn::Dropout2d conv2_drop;
	torch::nn::Linear fc1;
	torch::nn::Linear fc2;
};

cc @yf225 @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

**Comment by albanD** [comment]
Hi,

In the future, please use the forum to ask questions: https://discuss.pytorch.org/

I don't think there is any plan to support this from the onnx side.

**Comment by michaelboth** [comment]
> I don't think there is any plan to support this from the onnx side.

Hi albanD,

I'm not sure I understand your statement. Are you saying that libtorch won't support ONNX exports of model+weights? ONNX support seems so important, our company uses ONNX models to import into TensorRT. We are an embedded company that require absolute high performance, even with in-house training, and we want to use libtorch C++ to do training (avoiding python), then export the model+weights to ONNX, then import to NVIDIA's TensorRT.

Currently in PyTorch (python), you can:
```
dummy_input = torch.randn(1, 3, 224, 224, device=‘cuda’)
input_names = [ “input” ]
output_names = [ “output” ]
torch.onnx.export(model, dummy_input, “my_model.onnx”, verbose=True, input_names=input_names, output_names=output_names)
```

Does this mean a LibTorch C++ equivalent won't be implemented?

As an alternative, we tried training in C++, saving the model and weights using:
`torch::save(model, "model_and_weights.pt");`
Then tried to load this in PyTorch (python) with:
`model = torch.load("model_and_weights.pt")`
but this fails.

We also tried a pickle save on the nightly build of C++ side like so:
`std::vector<char> zip_data = torch::pickle_save(model);`
but that won't even compile, because it looks like pickle_save() can only take generic tensors, but not trained models.

Is there no way to train a model in C++ (avoiding python) and then export to ONNX (python or C++)?
Thank you for any help you can provide.

**Comment by yf225** [comment]
I think currently the easiest path is:

Define a JIT module in Python -> Train the JIT module in C++ -> Save the JIT module in C++ -> Load the JIT module in Python -> Export the JIT module to ONNX

For training the JIT module in C++, we might need to do something similar to https://github.com/pytorch/pytorch/issues/28478#issuecomment-574540294 to recursively get all parameters of a JIT module for training.

**Comment by michaelboth** [comment]
> Define a JIT module in Python -> Save the JIT module in Python -> Load the JIT module in C++ -> Train the JIT module in C++ -> Save the JIT module in C++ -> Load the JIT module in Python -> Export the JIT module to ONNX

Hi yf225,

By any chance, do you know of a good working example I can look at as a reference (maybe on GitHub)?

Thanks

**Comment by VitaMusic** [comment]
Could you clear the situation about convertion of libtorch models in onnx format ?
This option is critical for me, and I can not use the solution proposed by **michaelboth** via JIT in Python.
Whom and how to ask that this option be implemented in the near future? Thanks

**Comment by marcoippolito** [comment]
I join my voice to  @michaelboth and @VitaMusic 's  request to ask , if it is possible, to focus more effort on the C++ side of the already great PyTorch project.

There is a growing demand for implementing PyTorch's models directly in C++, without needing to jump back-and-forth between Python-C++, and again passing through ONNX.

Last November I asked in the PyTorch Forum's news about a timeline of the major breaking changes to the C++ backend: 
https://discuss.pytorch.org/t/c-api-is-there-a-timeline-of-the-major-breaking-changes-to-the-backend/60384/6

But, as far as I see and understand, the C++ API is still in "experimental" stage.

**Comment by jianyin2016** [comment]
#### Mark with a comment.

**Comment by shapovalovts** [comment]
Any update on this?

**Comment by hyc-xyz** [comment]
Like to know any update. Thanks.

**Comment by QuantScientist** [comment]
Same here.

**Comment by spandantiwari** [comment]
As @albanD mentioned above, there's no immediate plan to support this.

**Comment by tom-huntington** [comment]
I've been able to save a jit module from hasktorch, load it into PyTorch, and export it to onnx.

I've posted the code in this github issue https://github.com/hasktorch/hasktorch/issues/583

Here's the code where hasktorch calls libtorch that you would need to replicate
https://github.com/hasktorch/hasktorch/blob/a752f4c868caf386aeefafec9ab97e5810728f50/libtorch-ffi/src/Torch/Internal/Unmanaged/Type/Module.hs#L271

**Comment by garymm** [comment]
Duplicate of #17869.
